### DISCUSSION TITLE
Small question of workflow for a ground station

### DISCUSSION THREAD

**Discussion by billdion** [daniestevez/gr-satellites#319] [Q&A] [answered]
Hi Daniel,

I’m currently working on a ground station that should be able to decode incoming telemetry from passing satellites. I have hooked up Gpredict for the antenna control and have a doppler shift correction inside my GNURadio flowgraph. I’m using gr-satellite to try to print out the decoded PDU’s. I’m still a BIG noob in radio frequencies and still am learning each time I’m working on this project 😵‍💫. I wanted to ask you what should be my general flow to be capable of decoding telemetry. My current setup have a USRP that is multiplied with the doppler to make the frequency shift. After this I have the satellite decoder block. Here is an image to visualise it:

<img width="840" alt="Screen Shot 2021-11-06 at 3 29 41 PM" src="https://user-images.githubusercontent.com/31415925/140621778-220f6749-6cf3-4029-b07c-b1222788156b.png">

The cosine from the signal source is updated each 0.1sec via XMLRPC server! I DO see a beautiful red line coming out from the multiply block but I think that maybe a filter could help to extract it out from all the noise. I don’t see any data coming out from the satellite decoder block. I have tested out gr-satellite with some recordings from the submodule and I’m sure that they are working. I’m pretty sure the problem is coming from all the noise I’m getting from cranking the gain to 76 on the URSP.

I’m coming from a computer science background so I’m really more hands on with code than with RF theory. I wanted your thoughts on how could I extract something (I sure I’m missing some stuff), maybe it’s because of sample rate, but at this point I’m really clueless.

p.s. thanks for your work, I think I’m praising you at this point. 👍🏻

**Comment by daniestevez** [comment, marked as answer]
Hi. In general lines your decoding scheme looks good. If everything is working correctly you should get decodes with this flowgraph.

A suggestion is to tune the USRP away from the frequency of interest and use the Frequency Xlating FIR Filter to downconvert to baseband, filter and downsample the signal of interest. This avoids the DC frequency at the USRP, which can be problematic. Also, it's not necessary to run the satellite decoder at 192ksps, and depending on the USRP model, it might be advantageous to run it a sample rate higher than 192ksps, so decimation between the USRP and satellite decoder is advisable.

Besides this, I suggest you add a File Sink block so that you can record the the signal and take a look at it later on to assess signal quality and check for other problems. Looking at recordings with [inspectrum](https://github.com/miek/inspectrum) can be quite helpful.

Something else: you shouldn't use the "print" port of the Message Debug block (which prints out ASCII data), but rather the "print_pdu" port (which prints out a hex dump). Or you could use instead the Hexdump Sink from gr-satellites, which is just a wrapper around the print_pdu port of Message Debug.

**Reply by billdion** [reply]
Hi Daniel,

Thank you for this response. This helps a lot for my work! You made me realize that all this time I was looking at the DC frequency from the USRP. This was the straight red line I was seeing centered on my waterfall graph... this explains a lot 🤣. I'm usually outputting my signals on a ZMQ socket since I have gr-satellite dockerized with other components like a file sink to record satellite passes.

I've inputted 10M sample rate for the USRP with a centered frequency at 436.5M, I'm gonna let my software run all night to get good samples at ~62 gain into some files. I already checked your videos and I saw how you used the FIR filter to recenter the desired frequency of the satellite + a low pass to smooth things up.

Tomorrow I'm gonna use these files and I'm gonna tweak the center frequency on the FIR filters to try and capture data from those signals.


Many thanks,

p.s. I checked out quickly and I saw some "beeps" coming into the waterfall graph, I bet this contains more data than the DC line I saw before!

**Reply by daniestevez** [reply]
Good! I was intrigued by the "red line" you described, but didn't give it a second thought as to what could it be. It's possible to paste images here, so sometimes that helps.

**Reply by billdion** [reply]
I've arranged the USRP to 2.4M sample rate and for each satellite frequency I offset the centered frequency by 1M. Before going in the satellite decode block, the signal is passed inside a Frequency Xlating FIR Filter(complex band pass w/ 50 decimation) to downscale at 48khz (and center at 0)

This morning I got a hit for the Aistechsat-3 flyby;

![image](https://user-images.githubusercontent.com/31415925/140816110-b4527193-2de0-46d5-b63e-f4b1407047c9.png)

The decoded data looks like this: 

`83 D7 80 01 01 20 01 00 01 D0 3E 61 89 54 85 00 07 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 B9 30 29 AE`

This looks a lot like the frames I get with the example recordings for Aistechsat-3! So far it's the only satellite that gave me something but I'm going to keep looking up tonight for other hits. Gonna confirm if everything is working.

**Reply by daniestevez** [reply]
Seems a good decode. I'm a bit surprised that this packet is full of zeros, but who knows what AISTECHSAT-3 is sending these days (there's an open issue #85 about the content of its packets). Probably the last 4 bytes are a CRC-32C.

Not sure if it's an effect of the inspectrum power scale (which is set quite wide), but the signal doesn't look particularly strong. What kind of antenna are you using for this?

**Reply by billdion** [reply]
The antenna I'm using is a [UB-7030SAT](https://www.radioworld.ca/hyg-ub7030sat). I've boosted the gain to the maximum last night (76) to see if it will help, but I didn't get anything else than 2 more packets from Aistechsat. I am tracking a couple of satellites, in which Falconsat-3 is included (this one seems to spur out a lot of data).

 I'm going to check the recordings to see if it's just the signal coming in that is bad.

**Reply by billdion** [reply]
This is what came close to the 1Mhz offset

**Falconsat:**
![image](https://user-images.githubusercontent.com/31415925/140956790-bae181c4-3cf7-45fa-b163-65779ce15b51.png)

**Zhou Enlai:**
![image](https://user-images.githubusercontent.com/31415925/140956904-fac77b18-4ea3-4e75-b3c3-75159c05d85d.png)


Also, it seems that parsing back the file into the decoders gave me data, but it seems that it just picked up the noise (I got a LOT of data), here are some examples:

**Falconsat:**
- `60 E7 E8 79 C7 27 B3 05 54 0A A4 31 3D 82 8B CD 26 18 1F 4A 63 99 11 2C B0 C5`
- `48 13` (This one is small lol)
- `F8 C4 1F 8D D3 2D A3 DC D8 FF B4 52 81 87 04 D4 AA`

**Zhou Enlai:**
- `38 6E 9E 6B DE FB A2 63 77 2D 32 96 10 4D 1D 12 08 AE 56 71 51 FB D1 E0 A0 64 A3 F4 18 5B 97 04 2A 60 1C B2 07 A0 50 C8 B1 5E D8 72 E6 FE C9 84 8D 6F BD 88`
- `80 8F C9 A2 D9 79 50 39 F2 A6 3A 73 41 4E 26 A5 82 8D 86 02 0B 41 46 48 EA F5 A6 BE 3E 5C E9 4D 8F C7 BC 64 92 DA 9E E9 AD 60 29 2F 24 41 BE 97 78 DC 20 71 CE FD 01 B6 C9 82 5B DD`

This could very well be a hardware issue at this point.

**Reply by daniestevez** [reply]
Regarding adjusting the gain of the USRP, a good rule of thumb is that if you compare the noise floor coming from the antenna with the noise floor coming from a 50 Ohm load, the antenna noise should be at least 15 dB larger. Once this happens, the USRP noise figure is much lower than the noise temperature coming from the antenna, so you have set an adequate gain. Likewise, it isn't necessary to set the gain so high that you get a much larger difference when doing this test.

Your antenna has a lot of gain (for comparison, I often use a 7 element handheld on UHF), so you should definitely see strong signals. I'm assuming you have an LNA at the antenna. If not, unless you have a very short or very low loss run of coax to the USRP, you'll lose a lot of signal.

Regarding false decodes, some satellites (in particular those using AX.25) have weak CRC's, such as a CRC-16, so you'll get seemingly valid packets out of random noise. Not much to do to avoid this, other than set some heuristics based on the headers or setting up a power squelch.

**Reply by billdion** [reply]
This is good info. I bet that what's going on is the signals are too low to be decoded. Gonna take a look at hardware to tweak it! For now I think I can mark my question as answered 😄. Gonna post update if I can fix that low signal stuff.

Oh and, take a look at this packet from NETSAT4;

`00 00 CA 00 A6 68 60 88 A0 64 9C A6 88 61 03 F0 4E 7C 30 00 25 82 8D 06 07 01 06 01 02 3E 65 63 68 6F 20 4E 53 34 3E 42 3A 5C 31 5C 33 3B 52 54 43 47 45 54 49 3E 42 3A 5C 31 5C 33 3B 67 65 74 20 31 36 30 38 3E 42 3A 5C 31 5C 33 3B 67 65 74 20 31 36 30 39 3E 42 3A 5C 31 5C 33 97 16 14 F7 3F 75`

It translate to (without ax25 headers);

`>echo NS4>B:\1\3;RTCGETI>B:\1\3;get 1608>B:\1\3;get 1609>B:\1\3`

THATS GOOD! I like verbose data.

Thanks a lot!

**Comment by K4KDR** [comment]
If you would like some satellites that have particularly strong downlinks, a couple of good ones are:

RamSat  #48850  ( https://db.satnogs.org/satellite/YJWC-7364-7085-7563-0113#transmitters )

-and-

ELFIN-A  #43617  ( https://db.satnogs.org/satellite/JSEY-1831-4025-6206-5301#transmitters )

**Reply by billdion** [reply]
Oooooh thanks, that's great help! I'm adding them in my scheduler and we'll see how loud they talk 👀

**Reply by billdion** [reply]
ELFIN-A Gave me 1000 packets in a flyby, THIS IS INSANE! All of the packets seems to end with `5E` and start with `AE`, here are some examples;

- `AE 6C B2 A4 82 72 60 AE 94 64 B0 9C B0 E1 03 F0 93 95 44 01 10 21 00 30 04 01 A6 2B 04 D6 10 21 13 29 06 40 36 F1 00 E4 33 33 33 33 33 32 4C CD 27 93 33 33 33 33 36 4C CC C9 33 33 33 33 33 33 33 32 4C CC CC CC D9 33 33 64 D9 33 33 64 CC CC D9 33 64 C9 33 33 33 33 24 27 93 33 33 64 CC CC C9 33 33 32 4C CC CC CC CC CC CC 27 93 32 4D 27 93 33 33 33 33 33 33 33 36 15 86 66 66 66 66 66 66 64 B2 66 66 66 66 66 66 66 66 66 66 66 66 6C 99 99 B2 66 66 66 66 66 6C 99 99 99 99 99 99 99 B0 AC 33 24 CC 92 4C CC CC CC C9 32 4C CC CC CC CC CC CC 27 93 24 96 49 30 02 A0 05 F6 B0 5E`
- `AE 6C B2 A4 82 72 60 AE 94 64 B0 9C B0 E1 03 F0 93 95 B8 01 10 21 00 30 25 01 59 CA 04 4F 10 21 23 29 47 41 B2 3C 00 E5 33 24 CC C9 61 58 4B 0A C3 64 CC CC C9 33 33 36 4C CC C9 33 33 33 24 CC CC CC CC CC C9 64 D9 36 4C 27 93 25 92 49 32 4D 27 93 33 33 36 49 24 D9 33 33 33 64 92 15 84 B2 49 0A D2 15 08 49 0A C6 15 86 66 4B 26 CB 24 20 D8 56 30 AD 61 58 42 84 92 42 B0 CC CC 27 93 25 85 19 25 84 19 0A C2 10 C2 B1 85 61 26 C9 B2 49 B0 AC 32 48 56 19 B2 66 6C 99 99 B2 66 64 99 26 66 66 66 49 99 92 66 C9 99 99 9B 26 66 66 66 C2 B0 C9 33 24 CC CC C9 33 32 49 21 58 42 09 24 27 93 33 32 15 84 99 0A C6 15 86 C2 8C 28 C2 49 21 58 49 B2 6C 92 C2 86 42 B0 8C 96 15 A4 B0 86 36 48 56 12 42 B0 D9 24 CD 96 48 51 96 15 0D 96 4C D9 30 04 60 05 F6 60 5E`

I saw the data coming in live and I almost choked on my coffee

![image](https://user-images.githubusercontent.com/31415925/141137601-033827d1-c358-4d62-8f8d-5b4b77f57e7a.png)
